tokio-uds/src/frame.rs

```diff
@@ -0,0 +1,155 @@
+use std::io;
+use std::os::unix::net::SocketAddr;
+use std::path::PathBuf;
+
+use futures::{Async, Poll, Stream, Sink, StartSend, AsyncSink};
+
+use super::UnixDatagram;
+
+use tokio_codec::{Decoder, Encoder};
+use bytes::{BytesMut, BufMut};
+
+/// A unified `Stream` and `Sink` interface to an underlying `UnixDatagram`, using
+/// the `Encoder` and `Decoder` traits to encode and decode frames.
+///
+/// Unix datagram sockets work with datagrams, but higher-level code may wants to
+/// batch these into meaningful chunks, called "frames". This method layers
+/// framing on top of this socket by using the `Encoder` and `Decoder` traits to
+/// handle encoding and decoding of messages frames. Note that the incoming and
+/// outgoing frame types may be distinct.
+///
+/// This function returns a *single* object that is both `Stream` and `Sink`;
+/// grouping this into a single object is often useful for layering things which
+/// require both read and write access to the underlying object.
+///
+/// If you want to work more directly with the streams and sink, consider
+/// calling `split` on the `UnixDatagramFramed` returned by this method, which will break
+/// them into separate objects, allowing them to interact more easily.
+#[must_use = "sinks do nothing unless polled"]
+#[derive(Debug)]
+pub struct UnixDatagramFramed<C> {
+    socket: UnixDatagram,
+    codec: C,
+    rd: BytesMut,
+    wr: BytesMut,
+    out_addr: PathBuf,
+    flushed: bool,
+}
+
+impl<C: Decoder> Stream for UnixDatagramFramed<C> {
+    type Item = (C::Item, SocketAddr);
+    type Error = C::Error;
+
+    fn poll(&mut self) -> Poll<Option<(Self::Item)>, Self::Error> {
+        self.rd.reserve(INITIAL_RD_CAPACITY);
+
+        let (n, addr) = unsafe {
+            let (n, addr) = try_ready!(self.socket.poll_recv_from(self.rd.bytes_mut()));
+            self.rd.advance_mut(n);
+            (n, addr)
+        };
+        trace!("received {} bytes, decoding", n);
+        let frame_res = self.codec.decode(&mut self.rd);
+        self.rd.clear();
+        let frame = frame_res?;
+        let result = frame.map(|frame| (frame, addr));
+        trace!("frame decoded from buffer");
+        Ok(Async::Ready(result))
+    }
+}
+
+impl<C: Encoder> Sink for UnixDatagramFramed<C> {
+    type SinkItem = (C::Item, PathBuf);
```

Comment:
Is there a reason this takes a `PathBuf` instead of a `SocketAddr`?

Reply:
I can take another look, but I think you can only send to a datagram style socket if it has been bound to the filesystem somewhere. It needs to be SocketAddr on receive because it's possible to receive a message from an unbound socket.

Thinking about it now, I'm not sure how things work out with a connected pair of unbound sockets. Although that might not matter since we already can't use connected UDS sockets anyway.

Reply:
Looking again, it definitely needs to be this `PathBuf` because `SocketAddr` can't `AsRef` to `Path` (because it might not be) which is required by `UnixDatagram::send_to`.

Reply:
@carllerche, assuming you won't remember our gitter conversation by this point. 

If you're asking if this could use `AsRef<Path>` instead of `PathBuf` (because I don't have `AsRef<Path>` anywhere), yes it totally could. I have a change somewhere that I can push after work (several hours) if you'd rather have that. I can't recall the reason I went with an explicit `PathBuf`.

If you're asking about why it has to be some form of `Path`, it's because that what `UnixDatagram` already uses and (I assume) the rational behind that is that you can't send to an unconnected unnamed socket since this is connection-less, and `std::os::unix::net::SocketAddr` is either unnamed or a path.

Reply:
Ok, sorry for the delay...

Instead, maybe `UnixDatagramFramed` should be generic over the path type. `PathBuf` requires allocation. It would be unfortunate if using `UnixDatagramFramed` required allocation to send each frame if no allocation is required.

So, `UnixDatagramFramed` would be generic over the `T: AsRef<Path>` argument.

Reply:
@azdle it looks like this is the only piece of @carllerche's review feedback that hasn't yet been addressed. Are you interested in continuing to with this PR? If so, I'd be happy to approve it once you can make this change, otherwise, I'd be happy to wrap it up if you don't have the time.

Reply:
Oh, my bad, thought I had already replied here.

I have a commit I can push when I get home in a few hours that replaces `PathBuf` with `AsRef<Path>`. I can't remember why anymore, but I thought there was some reason that would cause more allocations.

I'll update again when I get a chance to look at that code again tonight.

Reply:
Ah, that sounds good. I also wrote an implementation in another branch that I think should _reduce_ the number of allocations --- when you push your commit, we can compare notes?

Reply:
@hawkw, didn't quite get to it last night, but here's a commit I could push here: https://github.com/azdle/tokio/commit/60085afb6285d31b777d243091547b9ae3b9b44c

I don't know what I was thinking with the [original code I had written awhile ago](https://github.com/azdle/tokio/commit/246eada7a066d7cbaded145c20df49d12b7c987c), but I definitely don't need the extra allocation that I had in there. With the update, I'm curios if you have thoughts on if there's a good way to avoid the allocation that's probably going to often needed in code that uses this API since it provides a `SocketAddr` and needs a `AsRef<Path>` back: https://github.com/azdle/tokio/commit/60085afb6285d31b777d243091547b9ae3b9b44c#diff-c246b64546cb315f4f405f0d3de05473R66